core/src/main/scala/cromwell/util/JsonEditor.scala

```diff
@@ -169,23 +206,75 @@ object JsonEditor {
+    * Update workflow json, replacing "subWorkflowMetadata" elements of root workflow's "calls" object by "subWorkflowId"
     *
     * @param workflowJson json blob
     * @return updated json
     */
-  def replaceSubworkflowMetadataWithId(workflowJson: Json): ErrorOr[Json] = {
-    def replaceSubworkflowMetadataObjectWithSubworkflowIdInCalls(callObject: JsonObject, subworkflowJson: Json): Option[ErrorOr[Json]] = {
-      Option(
+  def unexpandSubworkflows(workflowJson: Json): ErrorOr[Json] = {
+    import common.validation.ErrorOr._
+
+    def workflowObject(workflowJson: Json): ErrorOr[JsonObject] =
+      workflowJson.asObject.toErrorOr(s"Workflow JSON unexpectedly not a JSON object: $workflowJson")
+
+    def workflowObjectWithUpdatedCalls(workflowObject: JsonObject, callsObject: JsonObject): ErrorOr[Json] = {
+      def unexpandSubworkflow(callEntry: JsonObject, sub: Json): ErrorOr[Json] = {
         for {
-          subWorkflowId <- subworkflowJson.workflowId
-          updatedCallObj = callObject.remove(subWorkflowMetadataKey).add(subWorkflowIdKey, Json.fromString(subWorkflowId.toString))
-        } yield Json.fromJsonObject(updatedCallObj)
-      )
+          subObj <- sub.asObject.toErrorOr(s"subWorkflowMetadata unexpectedly not an object: $sub")
+          id <- subObj("id").toErrorOr(s"subWorkflowMetadata unexpectedly missing 'id' field: $subObj")
+          _ <- id.asString.toErrorOr(s"subworkflow 'id' field unexpectedly not a string: $id")
+          updated = callEntry.remove(subWorkflowMetadataKey).add(subWorkflowIdKey, id)
+        } yield Json.fromJsonObject(updated)
+      }
+
+      def processCallEntry(entry: Json): ErrorOr[Json] = {
+        entry.asObject match {
+          case None => s"call entry unexpectedly not an object: $entry".invalidNel
+          case Some(entryObject) =>
+            entryObject(subWorkflowMetadataKey) match {
+              case None => entry.validNel // no subworkflow metadata is fine, return the entry unmodified.
+              case Some(sub) => unexpandSubworkflow(entryObject, sub)
+            }
+        }
+      }
+
+      val updatedCallsObject = callsObject.traverse { json =>
+        val updatedCallArray: ErrorOr[Vector[Json]] = json.asArray match {
+          case Some(array) => array traverse processCallEntry
+          case None => s"value unexpectedly not an array: $json".invalidNel
+        }
+        updatedCallArray map Json.fromValues
+      }
+
+      for {
+        updatedCallsJson <- updatedCallsObject map Json.fromJsonObject
+      } yield Json.fromJsonObject(workflowObject.add("calls", updatedCallsJson))
     }
 
-    updateWorkflowCallsJson(workflowJson, replaceSubworkflowMetadataObjectWithSubworkflowIdInCalls)
+    for {
+      workflowObject <- workflowObject(workflowJson)
+      callsObject <- callsObject(workflowObject)
+      updated <- callsObject map { workflowObjectWithUpdatedCalls(workflowObject, _) } getOrElse workflowJson.validNel
```

Comment:
So if anything goes wrong anywhere in the process, we throw away the error message and return the original JSON?

Not saying that's _wrong_, just confirming?

Reply:
No error messages are thrown away, I'll add comments to clarify. 🙂 

`callsObject` is of type `Option[JsonObject]`, so the `map` has a return type of `Option[ErrorOr[Json]]`.